domain/miningmanager/mempool/mempool.go

```diff
@@ -406,32 +406,43 @@ func (mp *mempool) removeTransactionsFromPool(txs []*consensusexternalapi.Domain
 	return nil
 }
 
+type txAndOutpoint struct {
+	tx       *consensusexternalapi.DomainTransaction
+	outpoint *consensusexternalapi.DomainOutpoint
+}
+
 // removeTransactionAndItsChainedTransactions removes a transaction and all of its chained transaction from the mempool.
-//
 // This function MUST be called with the mempool lock held (for writes).
 func (mp *mempool) removeTransactionAndItsChainedTransactions(tx *consensusexternalapi.DomainTransaction) error {
-	txID := consensushashing.TransactionID(tx)
+	txAndOutpointQueue := make([]txAndOutpoint, 0)
+	txAndOutpointQueue = insertToTxAndOutpointQueue(txAndOutpointQueue, tx)
 	// Remove any transactions which rely on this one.
-	for i := uint32(0); i < uint32(len(tx.Outputs)); i++ {
-		prevOut := consensusexternalapi.DomainOutpoint{TransactionID: *txID, Index: i}
-		if txRedeemer, exists := mp.mempoolUTXOSet.poolTransactionBySpendingOutpoint(prevOut); exists {
-			err := mp.removeTransactionAndItsChainedTransactions(txRedeemer)
-			if err != nil {
-				return err
-			}
+	for len(txAndOutpointQueue) > 0 {
+		txnAndOutpoint := txAndOutpointQueue[0]
+		txAndOutpointQueue = txAndOutpointQueue[1:]
+		if txRedeemer, exists := mp.mempoolUTXOSet.poolTransactionBySpendingOutpoint(*txnAndOutpoint.outpoint); exists {
+			txAndOutpointQueue = insertToTxAndOutpointQueue(txAndOutpointQueue, txRedeemer)
 		}
-	}
 
-	if _, exists := mp.chainedTransactions[*txID]; exists {
-		mp.removeChainTransaction(tx)
+		txID := txnAndOutpoint.outpoint.TransactionID
+		if _, exists := mp.chainedTransactions[txID]; exists {
+			mp.removeChainTransaction(txnAndOutpoint.tx)
+		}
+		err := mp.cleanTransactionFromSets(txnAndOutpoint.tx)
+		if err != nil {
+			return err
+		}
 	}
+	return nil
+}
 
-	err := mp.cleanTransactionFromSets(tx)
-	if err != nil {
-		return err
+func insertToTxAndOutpointQueue(queue []txAndOutpoint, tx *consensusexternalapi.DomainTransaction) []txAndOutpoint {
```

Comment:
Is there a reason for `queue` to be a slice of values, as opposed to a slice of pointers?
Note that every time you e.g. append to this slice, a copy operation will take place.